src/main/java/uk/ac/soton/itinnovation/security/modelvalidator/Validator.java

```diff
@@ -2255,7 +2255,20 @@ private void calculateRelationCardinalityConstraint(CardinalityConstraintDB link
         // if the domain model doesn't have one, we use a default scale based on the asset's max
         // cardinality.
         String sourcePopulation = fromAsset.getPopulation();
+        if(sourcePopulation == null) {
+            logger.info("Found null population for asset {}", fromAsset.getUri());
+        }
+        if(poLevels.get(sourcePopulation) == null) {
+            logger.info("Found bad population {} for asset {}", sourcePopulation, fromAsset.getUri());
+        }
         String targetPopulation = toAsset.getPopulation();
+        if(targetPopulation == null) {
+            logger.info("Found null population for asset {}", toAsset.getUri());            
+        }
+        if(poLevels.get(targetPopulation) == null) {
+            logger.info("Found bad population {} for asset {}", targetPopulation, toAsset.getUri());
+        }
+
 
```

Comment:
Although these lines are slightly more defensive, I don't see that they prevent the original null pointer exception, as they are just logging issues. At the very least, it would be better to use "warn" instead "info", but it would be better to add some checks to the original line(s) that were failing.

Reply:
Sorry - I thought I removed those warnings. The fix is implemented in JenaQuerierDB.repairAssertedAssetPopulations().

The original purpose of this method was to convert 'pre-population' system models where assets would have a min/max cardinality instead of a population level. The method uses the max cardinality to select a population level.

This wasn't enough to handle the cut-and-paste bug, because that did insert a population level - but used a faulty URI. So the repair method was not triggered, leading to the crash. I added the warnings mainly so there was a line executed if the crash was about to happen, allowing me to add a break point and inspect the model state.

The fix in JenaQuerierDB.repairAssertedAssetPopulations() just checks if the population level is in the domain model scale and if not it overwrites the bad level. The only problem then is that because the asset was not created in an older SSM, there will be no max cardinality, so I added default levels to the AssetDB constructor method as well.

In principle, that should mean those warning messages in the validator can never appear. I just forgot to delete those lines.

Reply:
OK - looks fine now